### PR TITLE
Refactor config

### PR DESCRIPTION
**Pull request by Lee-W**
* Refactor config
    * Create a `BaseConfig` class for the default configuration
    * Use `TomlConfig` and `IniConfig` to separate the logic of parsing different type of config file
* Use new config structure to fix the previously fixed Issue #97
* Rename the config inside a config object as settings
* Use config class as an argument for **cz** and **commands** instead of dict config

### CODE REVIEW

**Comment by Woile** [comment]
Wow, great work Lee, I will review it this week

File: commitizen/config/ini_config.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import configparser
+import json
+import warnings
+
+from .base_config import BaseConfig
+
+
+class IniConfig(BaseConfig):
+    warnings.simplefilter("always", DeprecationWarning)
```
Comment (Woile) [inline commitizen/config/ini_config.py]: This is being executed always, even when a user is using `toml`, which is confusing. I'd move it to the `__init__`

File: commitizen/config/__init__.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import warnings
+from pathlib import Path
+
+from commitizen import defaults
+from .base_config import BaseConfig
+from .toml_config import TomlConfig
+from .ini_config import IniConfig
+
+
+def load_global_conf() -> dict:
```
Comment (Woile) [inline commitizen/config/__init__.py]: return `Optional[dict]`

File: commitizen/config/__init__.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    warnings.warn(message, category=DeprecationWarning)
+
+    with open(global_cfg, "r") as f:
+        data = f.read()
+
+    conf = IniConfig(data)
+    return conf
+
+
+def read_cfg() -> dict:
```
Comment (Woile) [inline commitizen/config/__init__.py]: This is also not dict. But I'm not sure which one is instead, I guess it's something like `typing.Type[BaseConfing]`

File: commitizen/config/toml_config.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        """
+        with open(self.path, "r") as f:
+            parser = parse(f.read())
+
+        parser["tool"]["commitizen"][key] = value
+        with open(self.path, "w") as f:
+            f.write(parser.as_string())
+        return self
+
+    def _parse_setting(self, data: str) -> dict:
```
Comment (Woile) [inline commitizen/config/toml_config.py]: This should either return something or remove the `-> dict`

File: commitizen/config/base_config.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+from commitizen.defaults import DEFAULT_SETTINGS
+
+
+class BaseConfig:
+    def __init__(self):
+        self._settings: dict = DEFAULT_SETTINGS.copy()
+        self._path: Optional[str] = None
+
+    @property
```
Comment (Woile) [inline commitizen/config/base_config.py]: If possible add the typehints of both properties. Thanks

File: commitizen/config/ini_config.py (hunk 1 above)
Comment (Lee-W) [reply commitizen/config/ini_config.py]: Yes, I added on another branch. I just cherry-pick it to this branch. Thanks!

File: commitizen/config/__init__.py (hunk 2 above)
Comment (Lee-W) [reply commitizen/config/__init__.py]: Done.

File: commitizen/config/toml_config.py (hunk 4 above)
Comment (Lee-W) [reply commitizen/config/toml_config.py]: Done.

File: commitizen/config/base_config.py (hunk 5 above)
Comment (Lee-W) [reply commitizen/config/base_config.py]: Done.

**Comment by Lee-W** [comment]
Maybe we can add mypy check in the future? But I think it might be better to do it after these PRs.

File: commitizen/config/__init__.py (hunk 3 above)
Comment (Lee-W) [reply commitizen/config/__init__.py]: Done.

**Pull request merged**